### ISSUE TITLE
Wrong hint message for `xi`

### ISSUE THREAD

**Issue by nullromo** [m4xshen/hardtime.nvim#77]
**Describe the bug**
In normal mode, pressing <kbd>x</kbd><kbd>i</kbd> will trigger a notification with the following message: `Use a instead of li`. Well, I didn't use `li` so this message makes no sense.

**Config options**
```lua
{ disable_mouse = false, restriction_mode = 'hint' }
```

**To Reproduce**
1. Open a file.
2. Press <kbd>x</kbd><kbd>i</kbd>.
3. Look at the message.

**Expected behavior**
The message should reflect the keys that were actually pressed.

**Additional context**
It looks to me like the source of this message is the following hint:
```lua
["[^fFtT]li"] = {
    message = function()
        return "Use a instead of li"
    end,
    length = 3,
}
```
However, I don't know how this could have triggered while using the <kbd>x</kbd> key.

**Other side notes**
An actual hint in this case would be `Use s instead of xi` because <kbd>s</kbd> does the same function as <kbd>x</kbd><kbd>i</kbd>. Maybe this should be added as a new hint... I should see if I can do this and submit a PR.

**Comment by nullromo** [comment]
I tried adding a custom hint like this

```lua
hints = {
    ['xi'] = {
        message = function()
            return 'Use s instead of xi'
        end,
        length = 2,
    },
},
```
but I can't get it to trigger. I tried with `length = 1`, `length = 2`, and `length = 3`.

Instead of getting my custom message, I just get the same old message. Additionally, I tried adding another hint:

```lua
hints = {
    ['xx'] = {
        message = function()
            return 'Do not use x multiple times consecutively'
        end,
        length = 2,
    },
},
```

but I cannot get this one to trigger either.

**Comment by kang8** [comment]
This is because [`x` same as `dl`](https://github.com/neovim/neovim/blob/1405e5c8c1f3b3ef59cab0ac2d0a28b9b88869cf/runtime/doc/change.txt#L21-L22). So pressing `x` is equivalent to pressing `dl`.

When you press `xi`, it is ultimately interpreted as `xdli`, which(`dli`) conforms to the `["[^fFtTlt]"` rule, so it matches.

You can use follow script to listener every input key

```lua
```

**Comment by nullromo** [comment]
It's weird that <kbd>x</kbd> actually "presses" <kbd>d</kbd><kbd>l</kbd> rather than just doing the underlying operation. For example, the following hint is kind of amusing:

```lua
['dl'] = {
    message = function()
        return 'Use x instead of dl'
    end,
    length = 2,
},
```

Each time you press <kbd>x</kbd>, it tells you to use <kbd>x</kbd> instead 🙃

So I guess hardtime.nvim can't actually distinguish which keypresses are "real" user keypresses. When I try your test function, I get the following notifications:

```
2024-02-02T09:15:03 test WARNING WARN WARNING | test: k is: x , key is: x
2024-02-02T09:15:03 test WARNING WARN WARNING | test: k is: d , key is: d
2024-02-02T09:15:03 hardtime WARNING WARN WARNING | hardtime: Use x instead of dl
2024-02-02T09:15:03 test WARNING WARN WARNING | test: k is: l , key is: l
```

This just feels like a very strange behavior of neovim to me.

**Comment by m4xshen** [comment]
@nullromo hardtime.nvim uses `vim.on_key` function to listen to key input. This behavior is expected based on the docs (see `:h vim.on_key`)

>         {fn} will receive the keys after mappings have been evaluated

I currently haven't found other way to listen to the **raw** key input.

**Comment by nullromo** [comment]
Looks like this is a feature missing from neovim that may (eventually) be added: https://github.com/neovim/neovim/issues/15527. 

In that case, feel free to close this issue as I feel it's a little too far out of the scope of hardtime at the moment.

**Comment by m4xshen** [comment]
> Looks like this is a feature missing from neovim that may (eventually) be added: https://github.com/neovim/neovim/issues/15527.

@nullromo Thanks for providing this information. I just update the code to use the latest api and now this issue is resolved! Btw the api requires Neovim >= 0.10
